src/main/scala/org/broadinstitute/hail/Utils.scala

```diff
@@ -1468,4 +1468,26 @@ object Utils extends Logging {
     }
     b.result()
   }
+
+  /**
+    * An abstraction for building an {@code Array} of known size. Guarantees a left-to-right traversal
+    *
+    * @param xs the thing to iterate over
+    * @param size the size of array to allocate
+    * @param key given the source value and its source index, yield the target index
+    * @param combine given the target value, the target index, the source value, and the source index, compute the new target value
+    * @tparam A
+    * @tparam B
+    */
+  def coalesce[A,B : ClassTag](xs: GenTraversableOnce[A])(size: Int, key: (A, Int) => Int, z: B)(combine: (B, Int, A, Int) => B): Array[B] = {
```

Comment:
I'm inclined to remove the indices.  You don't use them and if you want them you can always run `zipWithIndex` first.  That isn't true of the `B` index in `combine`, so I'm on the fence on that one, although I don't think it's needed in the common case.